### ISSUE TITLE
Google authentication is producing errors

### ISSUE THREAD

**Issue by joshkh** [intermine/intermine#1533]
As Paulo pointed out, logging into an existing mine using Google now reports a warning:

`Error granting permission: No two users may have the same value for 'alias', but the value '' is already taken`

... and users are left at the login screen after being authenticated. This is confusing.

In April of 2015 we provided functionality to migrate users from Google's defunct OpenID platform to their newer OpenID Connect platform. Similar names but the latter is actually oAuth2. See:

https://github.com/intermine/intermine/issues/854
https://github.com/intermine/intermine/pull/982

I have a feeling that Google has since changed the output of their OpenID Connect endpoint which is causing InterMine to attempt to migrate existing Google accounts, even though the existing user is no longer (or never was) an old OpenID account. Since the accounts have already been migrated InterMine thinks there are duplicates.

The Exception being thrown is here:
https://github.com/intermine/intermine/blob/dev/intermine/api/main/src/org/intermine/api/profile/DuplicateMappingException.java#L22

Which is being thrown from here:
https://github.com/intermine/intermine/blob/dev/intermine/api/main/src/org/intermine/api/profile/PreferencesManager.java#L237

**Comment by joshkh** [comment]
New Google accounts are being marked as "__migrated__:googleid" in the UserProfile table if they're new (I'm testing with an account less than a month old).

I believe it's a problem with our `!= NULL` checks versus empty strings. MigrationMapping has two properties: `oldId`, `newId`. While an old ID doesn't exist, I think `oldId` is returning an empty string, which passes our NULL checks here:

https://github.com/intermine/intermine/blob/dev/intermine/web/main/src/org/intermine/web/struts/oauth2/Callback.java#L314

https://github.com/intermine/intermine/blob/dev/intermine/web/main/src/org/intermine/web/struts/oauth2/Callback.java#L325

When I add a check for an empty string my username in the UserProfile table no longer contains "migrated".

**Comment by joshkh** [comment]
Also, we're storing empty strings in the UserPreferences table:

![image](https://cloud.githubusercontent.com/assets/4752532/25129568/8c373bc2-2436-11e7-9690-fd4ff26d76da.png)

**Comment by julie-sullivan** [comment]
What are the aliases used for? Oh, the "display name" on the webapp. That's why mine is blank now.  However it didn't used to be! What has google changed in their API to make this alias go away?

There seems like there may be two solutions:

1. Update to allow blank aliases. User won't have name next to log out button but won't get an error on log in!
2. Later, update our code to adapt to whatever changes Google made. Maybe if there is no alias, use email address?

**Comment by joshkh** [comment]
More notes:

Google always return an id_token for valid Google users:
```
[2017-04-18 13:02:47,530] ERROR org.intermine.web.struts.oauth2.Callback.googleProviderFlow(Callback.java:133) - GOOGLE RESPONSE: {
 "access_token": "JWT.Madeup.Token",
 "token_type": "Bearer",
 "expires_in": 3600,
 "id_token": "editedsgohusdghosudghsoudghsoudghsodug"
}
```
And what we get back from Google after exchanging token:

```
{"at_hash":"sgsdsgsdgsdg","aud":"sdgsdgsdgsdg-sdgsdgsdgsdg.apps.googleusercontent.com","sub":"28305834034686340468","email_verified":true,"azp":"ssdgsdgsdgsdg-sdgsdgsgsdg.apps.googleusercontent.com","iss":"accounts.google.com","exp":1492520567,"iat":39686039,"email":"joshkh@gmail.com"}
```

**Comment by julie-sullivan** [comment]
We are getting an error that the alias must be unique. 

We have lots of blanks, hence the duplicates.

But even if we didn't have blanks, we're going to have duplicate aliases.

Why do the aliases have to be unique?

**Comment by julie-sullivan** [comment]
```
2017-04-18 13:23:31 ERROR org.intermine.bio.dataconversion.EntrezPublicationsRetriever     - Couldn't parse PubMed XML
java.lang.NullPointerException
        at org.intermine.bio.dataconversion.EntrezPublicationsRetriever$FullRecordHandler.endElement(EntrezPublicationsRetriever.java:547)
        at org.apache.xerces.parsers.AbstractSAXParser.endElement(Unknown Source)
        at org.apache.xerces.impl.dtd.XMLDTDValidator.endNamespaceScope(Unknown Source)
        at org.apache.xerces.impl.dtd.XMLDTDValidator.handleEndElement(Unknown Source)
        at org.apache.xerces.impl.dtd.XMLDTDValidator.endElement(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanEndElement(Unknown Source)
```

Errors in the log. But it went ahead and processed other PubMed IDs.

**Comment by joshkh** [comment]
Okay, first things first, we need to add the 'profile' scope to global.web.properties:

https://github.com/intermine/intermine/commit/a7109773242ef968b2b02d7ac62f9884fd33648a

This allows us to see the names of new(ish) Google accounts.

**Comment by joshkh** [comment]
This is what I've gathered so far..

Users that are logged in via google are automatically given an 'alias' which is used for sharing lists. If an alias already exists with the same name then the user gets a " 1" appended to their name (this happens in Callback.java).

For testing, I'm putting break points in:

Callback.java - This is where the majority of the login procedure happens. You can also
Profile.java
UserPreferences.java

And looking for thrown "DuplicateMappingExceptions"

**Comment by joshkh** [comment]
Design decision moving forward for oAuth2 users:

1. Always write the user's email address to the userpreferences table. In theory this should never change, but the empty string fiasco revealed an edge case and this corrects it.

Then, if we have a valid identity from the provider (name, email address, etc)...

2. If the AKA field is never set, or it is set and blank, write it to the userpreferences table.

3. If the ALIAS field is never set, or it is set and blank, write it to the userpreferences table.

When a user logs in this triggers a rewrite to the preferences table that replaces any incorrect email addresses or blank strings.

**Comment by justinccdev** [comment]
Tested and updates #1541 to improve logic clarity and satisfy checkstyle.  Whilst testing, 2 things I noticed:

* Sometimes the first login silently fails and just dumps you back on the login screen.  Only tried this with Google logins.
* When sharing lists, the name displayed is something like "GOOGLE:105796464175245069917" instead of the user's proper name.  I don't believe this is directly related to this code change.  Maybe something to investigate afterwards.

I think we should still go ahead and commit these changes on Monday and address other issues subsequently.
